Feature/HMD Enhancements

FatLurch opened the pull request:
**When merged this pull request will:**
- Replaces #217 and addresses original comments
- Address most of #192 by significantly increasing the HMD brightness [Video](https://youtu.be/cLyzB-LyFrQ)
- Allows for user selectable HMD colors #180
- Allows for user selectable default HMD brightness
- Adds system to save and restore pylon config automatically to avert the "HMD fluid" issue [Video](https://youtu.be/H3gG6AKU83c)
- Harmonizes colors of HMD elements
- Groups HMD settings

FatLurch committed:
Initial Commit

FatLurch committed:
Delete settings.json

Local file. Don't want to pollute the repo

ampersand38 commented on addons/uh60_fms/config/interaction.hpp:
```diff
@@ -190,6 +190,12 @@ class hmd {
     buttonUp = "[4] call vtx_uh60_anvishud_fnc_toggleHud;";
   };
+  FMS_BTN(FMS_4,"BRT","")
+    buttonUp = [(_this select 0), ((driver vehicle player) isEqualTo player), 1] call vtx_uh60_anvishud_fnc_changeBrightness;
```
Seems like you don't need to pass the 2nd parameter, since the evaluation is always the same?

FatLurch replied:
This is how I determine if the player calling the function is pilot or copilot. In local testing it was giving me true for right seat and false for right seat, even after taking controls. 

I had set it up this way based on the feedback from #217.

ampersand38 replied:
right, but `player` is already unique where it's run, and I don't think we intend to change the other player's brightness. So it's not needed as a parameter and can just be evaluated inside `changeBrightness`.

FatLurch replied:
Gotcha, makes sense. Addressed in this commit:
https://github.com/Project-Hatchet/H-60/pull/247/commits/6c7f45809cd07a2f7eaa0fbf14eaaebc7414dd33

ampersand38 commented on addons/uh60_mfd/functions/fnc_perSecond.sqf:
```diff
@@ -8,6 +8,8 @@
 
 params ["_vehicle"];
 
+[_vehicle] call vtx_uh60_mfd_fnc_storePylons;
```
Maybe should be done in `changeBrightness` and `configureHMD` instead of perSecond?

FatLurch replied:
The intent is to store the configuration halfway often to manage cases like being reloaded by an ammo vehicle, etc. where the pilot won't be getting out of the aircraft.

I wanted to also catch the case where one or more of the EICAS messages could be asserted and it would be restored correctly.

In local testing this seemed to mitigate the "HMD fluid" issue.

ampersand38 replied:
We want to store the intentional changes, which as far as I can see are done by the brightness interactions and the color-from-settings. What other changes do we expect to need to store?

FatLurch replied:
I wanted to pickup the EICAS warnings as well which could assert or de-assert independent of an HMD action: 
https://docs.google.com/spreadsheets/d/1FUSzAR3CNno2-H-3twC4JH8l7Z7tsOUtHhMDkFxVoCM/edit#gid=0

Example: we could have a chip light come on due to damage. Pilot flies back for rearm (repair is unavailable) without interacting with the HMD or exiting the aircraft. The rearm action would "hide" the fact that we have a chip light on if we don't store that information at some frequency.

By the time the reloaded EH fires, the change to the pylon ammo quantity has already happened.

ampersand38 replied:
I see, yes, for everything that might go in the pylons. That speaks more to using a proper setter function to change any pylon value. PerSecond doesn't run when no pilot is in, so if they get out to repair and rearm, the stored values would be out of date.

FatLurch replied:
Ah, good point. I hadn't thought of instances where the pilot leaves the aircraft. 

I can come up with a function to update the pylon value and then store it, then work it into the existing pylon usage. It will take me a little while.

Thanks

FatLurch replied:
Initially addressed in commit: https://github.com/Project-Hatchet/H-60/pull/247/commits/ac5ca91f2bb031ef2ea8cec5328b69d3d29f8f38

ampersand38 commented on addons/uh60_mfd/functions/fnc_storePylons.sqf:
```diff
@@ -0,0 +1,21 @@
+{
+	private _index = _x;
+
+	if(_index select 3 == "vtx_1000rnd_dummy") then {
```
spaces =p

ampersand38 commented on addons/uh60_mfd/functions/fnc_storePylons.sqf:
```diff
@@ -0,0 +1,21 @@
+	};
+}forEach getAllPylonsInfo _vehicle;
+
+_vehicle setVariable ["pylonConfig", _pylonArray];
```
prefix the variable. `QGVAR(pylonConfig)`. if not prefixed any other mod might overwrite the variable.

ampersand38 commented on addons/uh60_mfd/functions/fnc_restorePylons.sqf:
```diff
@@ -0,0 +1,17 @@
+	private _index = _x;
+
+	_vehicle setPylonLoadout [_index select 0, "vtx_1000rnd_dummy", true];
+	_vehicle setAmmoOnPylon [_index select 0, _index select 1];
```
`[_index select 0, _index select 1]` this is just `_x` right?

FatLurch replied:
Correct. I got in the habit of putting _x into it's own variable due to nested forEach loops, but it's not strictly necessary here. I'll just use _x.

ampersand38 commented on addons/uh60_mfd/functions/fnc_storePylons.sqf:
```diff
@@ -0,0 +1,21 @@
+private _pylonArray = [];
+
+{
+	private _index = _x;
```
if this is for readability, consider
`_x params ["_index", "", "", "", "_value"];`

ampersand38 commented on addons/uh60_anvishud/initSettings.sqf:
```diff
@@ -11,3 +11,30 @@
+    "SLIDER",
+    [localize LSTRING(Brightness), LSTRING(Brightness_Description)],
+    ["UH-60M", "HMD"],
+    [0, 1, 0.6, 0, true],
```
Should probably be 2 decimal places instead of 0. Not sure what `true` is.

FatLurch replied:
Good catch on the decimals. The `true` expresses the value as a percentage in options (it's poorly documented, can't remember where I saw this originally)

FatLurch committed:
Addressing feedback

FatLurch committed:
Broke out params for _x

FatLurch committed:
Update in response to https://github.com/Project-Hatchet/H-60/pull/247#discussion_r776889795

FatLurch committed:
Address comment: https://github.com/Project-Hatchet/H-60/pull/247#discussion_r776888907

Initial incorporation of function to store pylon data in lieu of using perSecond.

Will expand to other instances where a pylon value is set in another commit

FatLurch committed:
All pylon values utilize setPylonValue

There were far fewer instances of setting pylon values than I had guessed
